src/pkcs11/framework-pkcs15.c

```diff
@@ -5310,7 +5310,16 @@ struct sc_pkcs11_object_ops pkcs15_profile_ops = {
 static void
 pkcs15_skey_release(void *object)
 {
-	__pkcs15_release_object((struct pkcs15_any_object *) object);
+	struct pkcs15_skey_object *skey = (struct pkcs15_skey_object *)object;
+	struct sc_pkcs15_skey_info *skey_info = skey->info;
+	struct sc_pkcs15_object *p15obj = skey->skey_p15obj;
+
+	if (__pkcs15_release_object((struct pkcs15_any_object *)skey) == 0) {
+		if (p15obj->session_object) {
```

Comment:
I really wonder why checking for session_object here is necessary. since skey_info is either NULL or created with calloc, which in both cases means that calling `free` on the pointer or the members of the data structure should be OK.

Reply:
I tried to describe this in the commit message. The `sc_pkcs15init_store_secret_key()` checks the `session_object` flag and if it is not set, it calls `sc_pkcs15init_add_object()`, which adds the p15 object to the directory structure that is freed during cleanup. Otherwise the structure is referenced only from the pkcs11 layer and need to be freed from that side.

The confusing thing with the skeys is that the skey_info

Looking at it more, we might also need some tweaks form the cards without onboard session objects (`SC_CARD_CAP_ONCARD_SESSION_OBJECTS` -- I think sc-hsm might be a bit different there and I probably did not check well).